<img width="200" style="float:left" 
     src="https://upload.wikimedia.org/wikipedia/commons/f/f3/Apache_Spark_logo.svg" />

<img style="float:left" src="https://storage.googleapis.com/kaggle-competitions/kaggle/3136/logos/header.png" />

# Sections
* [Description](#0)
* [1. Setup](#1)
  * [1.1 Start Hadoop](#1.1)  
  * [1.2 Search for Spark Installation](#1.2)
  * [1.3 Create SparkSession](#1.3)
* [2. Lab](#2)
  * [2.1 Check Lab Files](#2.1)
* [3. Data Preparation](#3)
  * [3.1 Data Cleansing](#3.1)
  * [3.2 Feature Engineering](#3.2)
* [4. Model Training](#4)
* [5. Model Evaluation](#5)
* [6. Model Selection](#6)
* [7. Model Persistence](#7)
* [8. Model Loading](#8)
* [9. TearDown](#9)
  * [9.1 Stop Hadoop](#9.1)

<a id='0'></a>
## Description
<p>
<div>The goal for this lab is:</div>
<ul>    
    <li>Practice Spark's Machine Learning API</li>
</ul>
</p>

The goal is to create a machine learning model to predict if a passenger would survive or not, therefore is a classification problem.

<a id='1'></a>
## 1. Setup

Since we are going to process data stored from HDFS let's start the service

<a id='1.1'></a>
### 1.1 Start Hadoop

Start Hadoop

Open a terminal and execute
```sh
hadoop-start.sh
```

In [75]:
import os

# Set Hadoop home directory (parent of 'bin' where winutils.exe is located)
os.environ["HADOOP_HOME"] = r"C:\hadoop"

# Add Hadoop bin directory (where winutils.exe is located) to PATH
os.environ["PATH"] += os.pathsep + r"C:\hadoop\bin"

# Verify if environment variables are set correctly
print("HADOOP_HOME:", os.environ.get("HADOOP_HOME"))
print("PATH:", os.environ.get("PATH"))

HADOOP_HOME: C:\hadoop
PATH: c:\Users\SLO\Documents\GitHub\IE-University\myenv312\Scripts;C:\Users\SLO\Documents\GitHub\IE-University\myenv312\Scripts;C:\Program Files\Python38\Scripts\;C:\Program Files\Python38\;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Program Files\dotnet\;C:\Program Files\Git\cmd;C:\Program Files (x86)\Microsoft SQL Server\160\Tools\Binn\;C:\Program Files\Microsoft SQL Server\160\Tools\Binn\;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn\;C:\Program Files\Microsoft SQL Server\160\DTS\Binn\;C:\Program Files\ibm\gsk8\lib64;C:\Program Files (x86)\ibm\gsk8\lib;C:\ProgramData\chocolatey\bin;C:\Users\SLO\AppData\Local\Microsoft\WindowsApps;C:\Users\SLO\AppData\Local\Programs\Microsoft VS Code\bin;C:\Users\SLO\AppData\Local\GitHubDesktop\bin;C:\Program Files\Java\jdk-11.0.26\bin;C:\spark\bin;C:\hadoop\bin;C:\Users\SLO\AppData\Local\Programs\Python\Python313;;C:\

In [76]:
!winutils.exe ls C:\

d--------- 1 NT SERVICE\TrustedInstaller NT SERVICE\TrustedInstaller 8192 Feb 25 2025 C:\


<a id='1.2'></a>
### 1.2 Search for Spark Installation 
This step is only needed in our course environment; other Spark environments you might see out there, might not need this statement.

In [77]:
import findspark
findspark.init()

In order to improve how data is displayed, I'll setup Pandas accordingly.

In [78]:
import pandas as pd
pd.set_option('display.max_colwidth', None)

<a id='1.3'></a>
### 1.3 Create SparkSession

Time to create the SparkSession which we'll use to send our Spark code:

In [79]:
from pyspark.sql.session import SparkSession

spark = (SparkSession.builder
    .appName("Titanic - Analytics - MLlib")
    .getOrCreate())

<a id='2'></a>
## 2. Lab

<a id='2.1'></a>
### 2.1 Check Lab Files

In order to complete this lab you need to previosly upload the datasets into HDFS.<br/>

Check you have the data ready in HDFS

http://localhost:50070/explorer.html#/datalake/raw/titanic/

<a id='2.2'></a>
### 2.2 Data Information
We have the following information about the dataset:
<table style="float:left">
<tbody>
<tr><th><b>Variable</b></th><th><b>Definition</b></th><th><b>Key</b></th></tr>
<tr>
<td>survival</td>
<td>Survival</td>
<td>0 = No, 1 = Yes</td>
</tr>
<tr>
<td>pclass</td>
<td>Ticket class</td>
<td>1 = 1st, 2 = 2nd, 3 = 3rd</td>
</tr>
<tr>
<td>sex</td>
<td>Sex</td>
<td></td>
</tr>
<tr>
<td>Age</td>
<td>Age in years</td>
<td></td>
</tr>
<tr>
<td>sibsp</td>
<td># of siblings / spouses aboard the Titanic</td>
<td></td>
</tr>
<tr>
<td>parch</td>
<td># of parents / children aboard the Titanic</td>
<td></td>
</tr>
<tr>
<td>ticket</td>
<td>Ticket number</td>
<td></td>
</tr>
<tr>
<td>fare</td>
<td>Passenger fare</td>
<td></td>
</tr>
<tr>
<td>cabin</td>
<td>Cabin number</td>
<td></td>
</tr>
<tr>
<td>embarked</td>
<td>Port of Embarkation</td>
<td>C = Cherbourg, Q = Queenstown, S = Southampton</td>
</tr>
</tbody>
</table>

<a id='3'></a>
## 3. Data Preparation

<a id='3.1'></a>
### 3.1 Data Cleansing

In [80]:
titanic_raw = (spark.read
                    .option("inferSchema", "true")
                    .option('header', 'true')
                    .csv(r"C:\Users\SLO\Documents\GitHub\IE-University\07_MODERN_DATA_ARCHITECTURES\titanic.csv")
                    .cache())

In [81]:
titanic_raw.limit(5).toPandas()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Thayer)  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500  None        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250  None        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500  None        S

In [82]:
titanic_raw.printSchema()

root
 |-- PassengerId: integer (nullable = true)
 |-- Survived: integer (nullable = true)
 |-- Pclass: integer (nullable = true)
 |-- Name: string (nullable = true)
 |-- Sex: string (nullable = true)
 |-- Age: double (nullable = true)
 |-- SibSp: integer (nullable = true)
 |-- Parch: integer (nullable = true)
 |-- Ticket: string (nullable = true)
 |-- Fare: double (nullable = true)
 |-- Cabin: string (nullable = true)
 |-- Embarked: string (nullable = true)



Let's do some *Exploratory Data Analysis* to understand our data a bit better

In [83]:
passengers_count = titanic_raw.count()
print (f"Total number of passenger: {passengers_count}")

Total number of passenger: 891


**Summary of data**:

In [84]:
titanic_raw.summary().toPandas()

summary        PassengerId             Survived              Pclass  \
0   count                891                  891                 891   
1    mean              446.0   0.3838383838383838   2.308641975308642   
2  stddev  257.3538420152301  0.48659245426485753  0.8360712409770491   
3     min                  1                    0                   1   
4     25%                223                    0                   2   
5     50%                446                    0                   3   
6     75%                669                    1                   3   
7     max                891                    1                   3   

                                               Name     Sex  \
0                                               891     891   
1                                              None    None   
2                                              None    None   
3  "Andersson, Mr. August Edvard (""Wennerstrom"")"  female   
4                                              None    None   
5                                              None    None   
6                                              None    None   
7                       van Melkebeke, Mr. Philemon    male   

                  Age               SibSp                Parch  \
0                 714                 891                  891   
1   29.69911764705882  0.5230078563411896  0.38159371492704824   
2  14.526497332334035  1.1027434322934315   0.8060572211299488   
3                0.42                   0                    0   
4                20.0                   0                    0   
5                28.0                   0                    0   
6                38.0                   1                    0   
7                80.0                   8                    6   

               Ticket               Fare Cabin Embarked  
0                 891                891   204      889  
1  260318.54916792738   32.2042079685746  None     None  
2  471609.26868834975  49.69342859718089  None     None  
3              110152                0.0   A10        C  
4             19996.0             7.8958  None     None  
5            236171.0            14.4542  None     None  
6            347743.0               31.0  None     None  
7           WE/P 5735           512.3292     T        S

Let's analize the number of **passengers who survived**:

In [85]:
titanic_raw.groupBy("Survived").count().toPandas()

Survived  count
0         1    342
1         0    549

342 out of the 891 passengers survived.

Let's dig into **specific information about survivors** by exploring some more data.

The survival rate can be determined by different features of the dataset such as *sex*, *port of embarcation*, *age*, ...

Let's analyze the **survival rate using feature sex**:

In [86]:
titanic_raw.groupBy("Sex","Survived").count().toPandas()

Sex  Survived  count
0    male         0    468
1  female         1    233
2  female         0     81
3    male         1    109

Even though the number of males is greater than the number of females in the ship, **females' survival rate is twice of the males'**.
Let's analyze the **survival rate using feature pclass**:

In [87]:
titanic_raw.groupBy("Pclass","Survived").count().toPandas()

Pclass  Survived  count
0       1         0     80
1       3         1    119
2       1         1    136
3       2         1     87
4       2         0     97
5       3         0    372

It's clear that people in *pclass #1* had much more priority than people in *pclass #3*; **even though the number of passengers in pclass #3 was higher, the survival rate was very low.**

In [88]:
titanic_df = titanic_raw

### Null Values

Let's **check if there are null values** that we need to remove before moving forward.

There are **two ways of coming up with the number of null values**:

In [89]:
from pyspark.sql.functions import isnull, when, count, col

# Option 1
titanic_df.select([count(when(isnull(c), c)).alias(c) for c in titanic_df.columns]).toPandas()

PassengerId  Survived  Pclass  Name  Sex  Age  SibSp  Parch  Ticket  Fare  \
0            0         0       0     0    0  177      0      0       0     0   

   Cabin  Embarked  
0    687         2

Another way of doing so:

In [90]:
# Option 2
titanic_df.summary().toPandas()

summary        PassengerId             Survived              Pclass  \
0   count                891                  891                 891   
1    mean              446.0   0.3838383838383838   2.308641975308642   
2  stddev  257.3538420152301  0.48659245426485753  0.8360712409770491   
3     min                  1                    0                   1   
4     25%                223                    0                   2   
5     50%                446                    0                   3   
6     75%                669                    1                   3   
7     max                891                    1                   3   

                                               Name     Sex  \
0                                               891     891   
1                                              None    None   
2                                              None    None   
3  "Andersson, Mr. August Edvard (""Wennerstrom"")"  female   
4                                              None    None   
5                                              None    None   
6                                              None    None   
7                       van Melkebeke, Mr. Philemon    male   

                  Age               SibSp                Parch  \
0                 714                 891                  891   
1   29.69911764705882  0.5230078563411896  0.38159371492704824   
2  14.526497332334035  1.1027434322934315   0.8060572211299488   
3                0.42                   0                    0   
4                20.0                   0                    0   
5                28.0                   0                    0   
6                38.0                   1                    0   
7                80.0                   8                    6   

               Ticket               Fare Cabin Embarked  
0                 891                891   204      889  
1  260318.54916792738   32.2042079685746  None     None  
2  471609.26868834975  49.69342859718089  None     None  
3              110152                0.0   A10        C  
4             19996.0             7.8958  None     None  
5            236171.0            14.4542  None     None  
6            347743.0               31.0  None     None  
7           WE/P 5735           512.3292     T        S

There are 3 features with missing data.

**Cabin** feature has 687 (891 - 204) null values.

**Embarked** feature has 2 (891 - 889) null values.

**Age** feature has 177 (891 - 714) null values.

Let's decide what to do

#### Cabin
Since there are so many missing values we have to get rid of this feature

In [91]:
titanic_df = titanic_df.drop("Cabin")

#### Embarked
Embarked feature has only two missining values. Let's check values within Embarked



In [92]:
titanic_df.groupBy("Embarked").count().toPandas()

Embarked  count
0        Q     77
1     None      2
2        C    168
3        S    644

Majority Passengers boarded from "S". We can impute these with "S"

In [93]:
titanic_df = titanic_df.na.fill({"Embarked" : 'S'})

#### Age
This is going to be a tricky one :) in some scenarios you can replace a missing value with the mean,median, mode value of the dataset.There is a functionality available is Spark for this job called <a href="https://spark.apache.org/docs/latest/ml-features.html#imputer">Imputer</a>

But this approach won't work here... **it might happen that you end up assignment a 4 years old kid a average age value of 29**.

We'll try to come up with a representative number based on other features. The **Name feature seems to be a good one** to calculate an average age for missing values.

**Names start with a word like Mr or Mrs**; calculating the average age for each group, will be closer to the real missing value... let's go for it:

In [94]:
from pyspark.sql.functions import regexp_extract
titanic_df = titanic_df.withColumn("Initial",regexp_extract(col("Name"),"([A-Za-z]+)\.",1))

<>:2: SyntaxWarning: invalid escape sequence '\.'
<>:2: SyntaxWarning: invalid escape sequence '\.'
C:\Users\SLO\AppData\Local\Temp\ipykernel_11388\1247204806.py:2: SyntaxWarning: invalid escape sequence '\.'
  titanic_df = titanic_df.withColumn("Initial",regexp_extract(col("Name"),"([A-Za-z]+)\.",1))


The regular expression "([A-Za-z]+)\\." extracts the words we're looking for: **text containing characters between A-Z or a-z and followed by a .(dot)**.

In [95]:
titanic_df.limit(5).toPandas()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Thayer)  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked Initial  
0      0         A/5 21171   7.2500        S      Mr  
1      0          PC 17599  71.2833        C     Mrs  
2      0  STON/O2. 3101282   7.9250        S    Miss  
3      0            113803  53.1000        S     Mrs  
4      0            373450   8.0500        S      Mr

In [96]:
titanic_df.select("Initial").distinct().sort("Initial").toPandas()

Initial
0       Capt
1        Col
2   Countess
3        Don
4         Dr
5   Jonkheer
6       Lady
7      Major
8     Master
9       Miss
10      Mlle
11       Mme
12        Mr
13       Mrs
14        Ms
15       Rev
16       Sir

There are some misspelled Initials like Mlle or Mme that stand for Miss. I will replace them with Miss and same thing for other values.

In [97]:
titanic_df = titanic_df.replace(
               ['Mlle','Mme', 'Ms', 'Dr','Major','Lady','Countess','Jonkheer','Col','Rev','Capt','Sir','Don'],
               ['Miss','Miss','Miss','Mr','Mr',  'Mrs',  'Mrs',  'Other',  'Other','Other','Mr','Mr','Mr'])

In [98]:
titanic_df.select("Initial").distinct().toPandas()

Initial
0    Miss
1   Other
2  Master
3      Mr
4     Mrs

Let's impute missing values in age feature based on average age of Initials
1. Calculate the average value based on the initials and create a DataFrame with those values:

In [99]:
from pyspark.sql.functions import round

avg_age_df = (titanic_df.groupby('Initial').avg('Age')
                        .withColumnRenamed("avg(Age)","Age"))
avg_age_df.toPandas()

Initial        Age
0    Miss  21.860000
1   Other  45.888889
2  Master   4.574167
3      Mr  32.739609
4     Mrs  35.981818

2. Let's create a temporary DF removing the Age column in those records/rows without a age (Age field with Null value):

In [100]:
titanic_df_noage = titanic_df.where(col("Age").isNull()).drop("Age")
titanic_df_noage.limit(1).toPandas()

PassengerId  Survived  Pclass              Name   Sex  SibSp  Parch  \
0            6         0       3  Moran, Mr. James  male      0      0   

   Ticket    Fare Embarked Initial  
0  330877  8.4583        Q      Mr

3. Add an Age field to the previous DataFrame by joining it with the average age DataFrame:

In [101]:
titanic_df_noage_with_avg = titanic_df_noage.join(avg_age_df, "Initial")
titanic_df_noage_with_avg.limit(1).toPandas()

Initial  PassengerId  Survived  Pclass              Name   Sex  SibSp  \
0      Mr            6         0       3  Moran, Mr. James  male      0   

   Parch  Ticket    Fare Embarked        Age  
0      0  330877  8.4583        Q  32.739609

4. Take the original titanic_df, **keep records/rows with a non-null age** and add the previous DataFrame to the result via an union transformation:

In [102]:
titanic_df_fixed = (titanic_df.where(col("Age").isNotNull())
                      .unionByName(titanic_df_noage_with_avg))

titanic_df_fixed.where(col("Age").isNull()).count()

0

In [103]:
titanic_df = titanic_df_fixed

#### Family_size and Alone
Let's create a couple of features called **Family_size** and **Alone**, which might bring some insights on survival rate and the size of the families.

**Family_size** is the total number of *parch (parents/children)* and *sibsp (siblings/spouses)* per row. **Alone** will be a flag set when the size of the family equals to 0.

In [104]:
titanic_df = titanic_df.withColumn("Family_Size",col('SibSp')+col('Parch'))

In [105]:
titanic_df.groupBy("Family_Size").count().toPandas()

Family_Size  count
0            1    161
1            6     12
2            3     29
3            5     22
4            4     15
5            7      6
6            2    102
7            0    537
8           10      7

In [106]:
from pyspark.sql.functions import lit
titanic_df = titanic_df.withColumn("Alone",when(titanic_df["Family_Size"] == 0, 1).otherwise(lit(0)))

In [107]:
titanic_df.columns

['PassengerId',
 'Survived',
 'Pclass',
 'Name',
 'Sex',
 'Age',
 'SibSp',
 'Parch',
 'Ticket',
 'Fare',
 'Embarked',
 'Initial',
 'Family_Size',
 'Alone']

<a id='3.2'></a>
### 3.2. Feature Engineering
It's time to convert data into a suitable format for machine learning algorithms.<br/> 
First let's get rid of columns with unique values that don't contribute to a persons survival probability

In [108]:
from pyspark.sql.functions import countDistinct

titanic_df.select([countDistinct(c).alias(c) for c in titanic_df.columns]).toPandas()

PassengerId  Survived  Pclass  Name  Sex  Age  SibSp  Parch  Ticket  Fare  \
0          891         2       3   891    2   92      7      7     681   248   

   Embarked  Initial  Family_Size  Alone  
0         3        5            9      2

In [109]:
titanic_df = titanic_df.drop("PassengerId","Name","Ticket","Initial")

In [110]:
titanic_df.printSchema()

root
 |-- Survived: integer (nullable = true)
 |-- Pclass: integer (nullable = true)
 |-- Sex: string (nullable = true)
 |-- Age: double (nullable = true)
 |-- SibSp: integer (nullable = true)
 |-- Parch: integer (nullable = true)
 |-- Fare: double (nullable = true)
 |-- Embarked: string (nullable = false)
 |-- Family_Size: integer (nullable = true)
 |-- Alone: integer (nullable = false)



Let's cast all numerical values to doubles

In [111]:
titanic_df = titanic_df.select(col('Survived').cast('double'),
                              col('Pclass').cast('double'),
                              col('Sex'),
                              col('Age').cast('double'),
                              col('SibSp').cast('double'),
                              col('Parch').cast('double'),
                              col('Fare').cast('double'),
                              col('Embarked'),
                              col('Family_Size').cast('double'),
                              col('Alone').cast('double')
                             )

In [112]:
titanic_df.printSchema()

root
 |-- Survived: double (nullable = true)
 |-- Pclass: double (nullable = true)
 |-- Sex: string (nullable = true)
 |-- Age: double (nullable = true)
 |-- SibSp: double (nullable = true)
 |-- Parch: double (nullable = true)
 |-- Fare: double (nullable = true)
 |-- Embarked: string (nullable = false)
 |-- Family_Size: double (nullable = true)
 |-- Alone: double (nullable = false)



#### 3.2.2 Feature Transformation

We need to **translate the values in string columns into  numerical values**.

In order to do so, we are going to *encode* categorical values using:<br/>

[StringIndexer](https://spark.apache.org/docs/latest/ml-features#stringindexer) <br/>
[OneHotEncoder](https://spark.apache.org/docs/latest/ml-features#onehotencoder)

In [113]:
label_column = "Survived"

categoricalCols = [field for (field, dataType) in titanic_df.dtypes if ((dataType == "string") & (field != label_column))]
numericCols = [field for (field, dataType) in titanic_df.dtypes if ((dataType == "double") & (field != label_column))]

print (f"categorical columns: {categoricalCols}")
print (f"numerical columns: {numericCols}")

categorical columns: ['Sex', 'Embarked']
numerical columns: ['Pclass', 'Age', 'SibSp', 'Parch', 'Fare', 'Family_Size', 'Alone']


In [114]:
indexOutputCols = [x + "Index" for x in categoricalCols]
oheOutputCols = [x + "OHE" for x in categoricalCols]

print (f"StringIndexer column names: {indexOutputCols}")
print (f"OHE column names: {oheOutputCols}")

StringIndexer column names: ['SexIndex', 'EmbarkedIndex']
OHE column names: ['SexOHE', 'EmbarkedOHE']


In [115]:
from pyspark.ml.feature import OneHotEncoder, StringIndexer

stringIndexer = StringIndexer(inputCols=categoricalCols, outputCols=indexOutputCols, handleInvalid="skip")

oheEncoder = OneHotEncoder(inputCols=indexOutputCols,outputCols=oheOutputCols)

Checking StringIndexer

In [116]:
temp_df = stringIndexer.fit(titanic_df).transform(titanic_df)
temp_df.toPandas()

Survived  Pclass     Sex        Age  SibSp  Parch     Fare Embarked  \
0         0.0     3.0    male  22.000000    1.0    0.0   7.2500        S   
1         1.0     1.0  female  38.000000    1.0    0.0  71.2833        C   
2         1.0     3.0  female  26.000000    0.0    0.0   7.9250        S   
3         1.0     1.0  female  35.000000    1.0    0.0  53.1000        S   
4         0.0     3.0    male  35.000000    0.0    0.0   8.0500        S   
..        ...     ...     ...        ...    ...    ...      ...      ...   
886       0.0     3.0    male  32.739609    0.0    0.0   7.2292        C   
887       0.0     3.0  female  21.860000    8.0    2.0  69.5500        S   
888       0.0     3.0    male  32.739609    0.0    0.0   9.5000        S   
889       0.0     3.0    male  32.739609    0.0    0.0   7.8958        S   
890       0.0     3.0  female  21.860000    1.0    2.0  23.4500        S   

     Family_Size  Alone  SexIndex  EmbarkedIndex  
0            1.0    0.0       0.0            0.0  
1            1.0    0.0       1.0            1.0  
2            0.0    1.0       1.0            0.0  
3            1.0    0.0       1.0            0.0  
4            0.0    1.0       0.0            0.0  
..           ...    ...       ...            ...  
886          0.0    1.0       0.0            1.0  
887         10.0    0.0       1.0            0.0  
888          0.0    1.0       0.0            0.0  
889          0.0    1.0       0.0            0.0  
890          3.0    0.0       1.0            0.0  

[891 rows x 12 columns]

Checking OneHotEncoder

In [117]:
oheEncoder.fit(temp_df).transform(temp_df).toPandas()

Survived  Pclass     Sex        Age  SibSp  Parch     Fare Embarked  \
0         0.0     3.0    male  22.000000    1.0    0.0   7.2500        S   
1         1.0     1.0  female  38.000000    1.0    0.0  71.2833        C   
2         1.0     3.0  female  26.000000    0.0    0.0   7.9250        S   
3         1.0     1.0  female  35.000000    1.0    0.0  53.1000        S   
4         0.0     3.0    male  35.000000    0.0    0.0   8.0500        S   
..        ...     ...     ...        ...    ...    ...      ...      ...   
886       0.0     3.0    male  32.739609    0.0    0.0   7.2292        C   
887       0.0     3.0  female  21.860000    8.0    2.0  69.5500        S   
888       0.0     3.0    male  32.739609    0.0    0.0   9.5000        S   
889       0.0     3.0    male  32.739609    0.0    0.0   7.8958        S   
890       0.0     3.0  female  21.860000    1.0    2.0  23.4500        S   

     Family_Size  Alone  SexIndex  EmbarkedIndex SexOHE EmbarkedOHE  
0            1.0    0.0       0.0            0.0  (1.0)  (1.0, 0.0)  
1            1.0    0.0       1.0            1.0  (0.0)  (0.0, 1.0)  
2            0.0    1.0       1.0            0.0  (0.0)  (1.0, 0.0)  
3            1.0    0.0       1.0            0.0  (0.0)  (1.0, 0.0)  
4            0.0    1.0       0.0            0.0  (1.0)  (1.0, 0.0)  
..           ...    ...       ...            ...    ...         ...  
886          0.0    1.0       0.0            1.0  (1.0)  (0.0, 1.0)  
887         10.0    0.0       1.0            0.0  (0.0)  (1.0, 0.0)  
888          0.0    1.0       0.0            0.0  (1.0)  (1.0, 0.0)  
889          0.0    1.0       0.0            0.0  (1.0)  (1.0, 0.0)  
890          3.0    0.0       1.0            0.0  (0.0)  (1.0, 0.0)  

[891 rows x 14 columns]

In [118]:
assemblerInputs = oheOutputCols + numericCols
print("Feature columns: ",assemblerInputs)

Feature columns:  ['SexOHE', 'EmbarkedOHE', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare', 'Family_Size', 'Alone']


#### 3.2.3 Feature Assembling

It's finally time to **assemble the features in one single vector**, which is what the algorithm will expect, by using something called [VectorAssembler](https://spark.apache.org/docs/latest/api/python/pyspark.ml.html?highlight=vectorassembler#pyspark.ml.feature.VectorAssembler).

As the **"Survived" variable** is the one we want to predict, **all the other variables** will be considered to build the **list with required features**:

In [119]:
from pyspark.ml.feature import VectorAssembler

vecAssembler = VectorAssembler(inputCols=assemblerInputs,outputCol="features")

In [120]:
from pyspark.ml import Pipeline

test_pipeline = Pipeline(stages = [stringIndexer, oheEncoder, vecAssembler])
features_df = test_pipeline.fit(titanic_df).transform(titanic_df)
features_df.limit(2).toPandas()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked  \
0       0.0     3.0    male  22.0    1.0    0.0   7.2500        S   
1       1.0     1.0  female  38.0    1.0    0.0  71.2833        C   

   Family_Size  Alone  SexIndex  EmbarkedIndex SexOHE EmbarkedOHE  \
0          1.0    0.0       0.0            0.0  (1.0)  (1.0, 0.0)   
1          1.0    0.0       1.0            1.0  (0.0)  (0.0, 1.0)   

                                                  features  
0     [1.0, 1.0, 0.0, 3.0, 22.0, 1.0, 0.0, 7.25, 1.0, 0.0]  
1  [0.0, 0.0, 1.0, 1.0, 38.0, 1.0, 0.0, 71.2833, 1.0, 0.0]

<a id='4'></a>
## 4. Model Training

Here is the list of few classification algorithms from Spark ML we are going to try:

<ul>
<li>LogisticRegression</li>
<li>DecisionTreeClassifier</li>
<li>RandomForestClassifier</li>
<li>Gradient-boosted tree classifier</li>
<li>NaiveBayes</li>
<li>Linear Support Vector Machine</li>
</ul>

There are some points in the machine learning workflow were randomness takes place, for example during the sets splitting and some ML algorithms like RandomForest.

In order to make our experiments reproducible and get always the same results with the same data no matter how many times we execute our code, we are going to use a seed.

The seed is tipically a prime number.

In [121]:
seed=11

In [122]:
from pyspark.ml.classification import LogisticRegression
from pyspark.ml.classification import DecisionTreeClassifier
from pyspark.ml.classification import RandomForestClassifier
from pyspark.ml.classification import GBTClassifier
from pyspark.ml.classification import NaiveBayes
from pyspark.ml.classification import LinearSVC

lr = LogisticRegression(labelCol="Survived", featuresCol="features")
dt = DecisionTreeClassifier(labelCol="Survived", featuresCol="features",seed=seed)
rf = RandomForestClassifier(labelCol="Survived", featuresCol="features",maxDepth=10,seed=seed)
gbt = GBTClassifier(labelCol="Survived", featuresCol="features",maxIter=10,seed=seed)
nb = NaiveBayes(labelCol="Survived", featuresCol="features")
svm = LinearSVC(labelCol="Survived", featuresCol="features")

classifiers = [lr,dt,rf,gbt,nb,svm]
classifiers

[LogisticRegression_d5b5b9b68a57,
 DecisionTreeClassifier_2309e0d12244,
 RandomForestClassifier_0b45bb31d614,
 GBTClassifier_4bc6d1ed3a07,
 NaiveBayes_8d5e2d88349a,
 LinearSVC_e946d7f4ba72]

Let's create a pipeline for every classifier

In [123]:
from pyspark.ml import Pipeline

def create_pipeline(classifier):
    return Pipeline(stages = [stringIndexer, oheEncoder, vecAssembler, classifier])

pipelines = [create_pipeline(classifier) for classifier in classifiers]
pipelines

[Pipeline_8ca7b7de140b,
 Pipeline_9e048825ffc0,
 Pipeline_e5bcac10f27c,
 Pipeline_3774f8a356b6,
 Pipeline_dd40e76a5d7d,
 Pipeline_d35ea190f22a]

<a id='5'></a>
## 5. Model Evaluation
We're going to evaluate our classification model by using [MulticlassClassificationEvaluator](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.ml.evaluation.MulticlassClassificationEvaluator.html)

We're going to use the accuray metric.

In [124]:
from pyspark.ml.evaluation import MulticlassClassificationEvaluator

evaluator = MulticlassClassificationEvaluator(labelCol="Survived",  metricName="accuracy")

<a id='6'></a>
## 6. Model Selection

Now that the data is all set, let's split it into training and test. We can use a 80-20 ratio

In [125]:
(trainingData, testData) = titanic_df.randomSplit([0.8,0.2],seed=seed)

It's a good practice to keep the same distribution of 0's and 1's in the training set, and specially critical in unbalanced/skew datasets. This is called **Stratified Train-Test Split**

In [126]:
(trainingData1, testData1) = titanic_df.where("Survived=0").randomSplit([0.8,0.2],seed=seed)
(trainingData2, testData2) = titanic_df.where("Survived=1").randomSplit([0.8,0.2],seed=seed)

traininData = trainingData1.unionByName(trainingData2)
testData = testData1.unionByName(testData2)

Let's train all the classifiers

In [127]:
models = [pipeline.fit(trainingData) for pipeline in pipelines]
models

[PipelineModel_1e304f6d466f,
 PipelineModel_7ad1ca2379ca,
 PipelineModel_0e7cf47b9f7a,
 PipelineModel_1e209676e4ed,
 PipelineModel_3e89ea63f525,
 PipelineModel_4b68ec8c3869]

Let's evaluate all the models

In [128]:
names = []
values = [] 
for model in models:
    prediction_df = model.transform(testData)
    accuracy = evaluator.evaluate(prediction_df)
    names.append(type(model.stages[-1]).__name__) # the algorithm is the last stage in the pipeline
    values.append(accuracy)

data = {'name':names,'accuracy':values,'model':models}
df = pd.DataFrame(data)
df.sort_values(by=['accuracy'], inplace=True, ascending=False)  
df

name  accuracy                       model
2  RandomForestClassificationModel  0.842391  PipelineModel_0e7cf47b9f7a
1  DecisionTreeClassificationModel  0.826087  PipelineModel_7ad1ca2379ca
5                   LinearSVCModel  0.826087  PipelineModel_4b68ec8c3869
3           GBTClassificationModel  0.820652  PipelineModel_1e209676e4ed
0          LogisticRegressionModel  0.804348  PipelineModel_1e304f6d466f
4                  NaiveBayesModel  0.701087  PipelineModel_3e89ea63f525

The best model is **RandomForestClassificationModel**

In [129]:
best_model=df.iloc[0]['model']

Confusion matrix

In [130]:
best_model.transform(testData).groupby("Survived").pivot("prediction").count().toPandas()

Survived  0.0  1.0
0       0.0   98   17
1       1.0   12   57

<a id='7'></a>
## 7. Model Persistence
Spark provides functionality to save the model/pipeline so that we can use it later for inference (batch or streaming)

In [134]:
import os

# Convert relative path to absolute path
modelPath = os.path.abspath("../MODELS/titanic_best_model")

print("Saving model to:", modelPath)

# Save the model
best_model.write().overwrite().save(modelPath)

Saving model to: c:\Users\SLO\Documents\GitHub\IE-University\07_MODERN_DATA_ARCHITECTURES\MODELS\titanic_best_model


Py4JJavaError: An error occurred while calling o6370.save.
: org.apache.spark.SparkException: Job aborted.
	at org.apache.spark.internal.io.SparkHadoopWriter$.write(SparkHadoopWriter.scala:106)
	at org.apache.spark.rdd.PairRDDFunctions.$anonfun$saveAsHadoopDataset$1(PairRDDFunctions.scala:1091)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:410)
	at org.apache.spark.rdd.PairRDDFunctions.saveAsHadoopDataset(PairRDDFunctions.scala:1089)
	at org.apache.spark.rdd.PairRDDFunctions.$anonfun$saveAsHadoopFile$4(PairRDDFunctions.scala:1062)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:410)
	at org.apache.spark.rdd.PairRDDFunctions.saveAsHadoopFile(PairRDDFunctions.scala:1027)
	at org.apache.spark.rdd.PairRDDFunctions.$anonfun$saveAsHadoopFile$3(PairRDDFunctions.scala:1009)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:410)
	at org.apache.spark.rdd.PairRDDFunctions.saveAsHadoopFile(PairRDDFunctions.scala:1008)
	at org.apache.spark.rdd.PairRDDFunctions.$anonfun$saveAsHadoopFile$2(PairRDDFunctions.scala:965)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:410)
	at org.apache.spark.rdd.PairRDDFunctions.saveAsHadoopFile(PairRDDFunctions.scala:963)
	at org.apache.spark.rdd.RDD.$anonfun$saveAsTextFile$2(RDD.scala:1623)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:410)
	at org.apache.spark.rdd.RDD.saveAsTextFile(RDD.scala:1623)
	at org.apache.spark.rdd.RDD.$anonfun$saveAsTextFile$1(RDD.scala:1609)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:410)
	at org.apache.spark.rdd.RDD.saveAsTextFile(RDD.scala:1609)
	at org.apache.spark.ml.util.DefaultParamsWriter$.saveMetadata(ReadWrite.scala:413)
	at org.apache.spark.ml.Pipeline$SharedReadWrite$.$anonfun$saveImpl$1(Pipeline.scala:250)
	at org.apache.spark.ml.Pipeline$SharedReadWrite$.$anonfun$saveImpl$1$adapted(Pipeline.scala:247)
	at org.apache.spark.ml.util.Instrumentation$.$anonfun$instrumented$1(Instrumentation.scala:191)
	at scala.util.Try$.apply(Try.scala:213)
	at org.apache.spark.ml.util.Instrumentation$.instrumented(Instrumentation.scala:191)
	at org.apache.spark.ml.Pipeline$SharedReadWrite$.saveImpl(Pipeline.scala:247)
	at org.apache.spark.ml.PipelineModel$PipelineModelWriter.saveImpl(Pipeline.scala:346)
	at org.apache.spark.ml.util.MLWriter.save(ReadWrite.scala:168)
	at org.apache.spark.ml.PipelineModel$PipelineModelWriter.super$save(Pipeline.scala:344)
	at org.apache.spark.ml.PipelineModel$PipelineModelWriter.$anonfun$save$4(Pipeline.scala:344)
	at org.apache.spark.ml.MLEvents.withSaveInstanceEvent(events.scala:174)
	at org.apache.spark.ml.MLEvents.withSaveInstanceEvent$(events.scala:169)
	at org.apache.spark.ml.util.Instrumentation.withSaveInstanceEvent(Instrumentation.scala:42)
	at org.apache.spark.ml.PipelineModel$PipelineModelWriter.$anonfun$save$3(Pipeline.scala:344)
	at org.apache.spark.ml.PipelineModel$PipelineModelWriter.$anonfun$save$3$adapted(Pipeline.scala:344)
	at org.apache.spark.ml.util.Instrumentation$.$anonfun$instrumented$1(Instrumentation.scala:191)
	at scala.util.Try$.apply(Try.scala:213)
	at org.apache.spark.ml.util.Instrumentation$.instrumented(Instrumentation.scala:191)
	at org.apache.spark.ml.PipelineModel$PipelineModelWriter.save(Pipeline.scala:344)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:374)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
	at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.UnsatisfiedLinkError: 'boolean org.apache.hadoop.io.nativeio.NativeIO$Windows.access0(java.lang.String, int)'
	at org.apache.hadoop.io.nativeio.NativeIO$Windows.access0(Native Method)
	at org.apache.hadoop.io.nativeio.NativeIO$Windows.access(NativeIO.java:793)
	at org.apache.hadoop.fs.FileUtil.canRead(FileUtil.java:1249)
	at org.apache.hadoop.fs.FileUtil.list(FileUtil.java:1454)
	at org.apache.hadoop.fs.RawLocalFileSystem.listStatus(RawLocalFileSystem.java:601)
	at org.apache.hadoop.fs.FileSystem.listStatus(FileSystem.java:1972)
	at org.apache.hadoop.fs.FileSystem.listStatus(FileSystem.java:2014)
	at org.apache.hadoop.fs.ChecksumFileSystem.listStatus(ChecksumFileSystem.java:761)
	at org.apache.hadoop.fs.FileSystem.listStatus(FileSystem.java:1972)
	at org.apache.hadoop.fs.FileSystem.listStatus(FileSystem.java:2014)
	at org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter.getAllCommittedTaskPaths(FileOutputCommitter.java:334)
	at org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter.commitJobInternal(FileOutputCommitter.java:404)
	at org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter.commitJob(FileOutputCommitter.java:377)
	at org.apache.hadoop.mapred.FileOutputCommitter.commitJob(FileOutputCommitter.java:136)
	at org.apache.hadoop.mapred.OutputCommitter.commitJob(OutputCommitter.java:291)
	at org.apache.spark.internal.io.HadoopMapReduceCommitProtocol.commitJob(HadoopMapReduceCommitProtocol.scala:192)
	at org.apache.spark.internal.io.SparkHadoopWriter$.$anonfun$write$3(SparkHadoopWriter.scala:100)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.util.Utils$.timeTakenMs(Utils.scala:552)
	at org.apache.spark.internal.io.SparkHadoopWriter$.write(SparkHadoopWriter.scala:100)
	... 68 more


In [133]:
modelPath = "../MODELS/titanic_best_model"
best_model.write().overwrite().save(modelPath)

Py4JJavaError: An error occurred while calling o6258.save.
: org.apache.spark.SparkException: Job aborted.
	at org.apache.spark.internal.io.SparkHadoopWriter$.write(SparkHadoopWriter.scala:106)
	at org.apache.spark.rdd.PairRDDFunctions.$anonfun$saveAsHadoopDataset$1(PairRDDFunctions.scala:1091)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:410)
	at org.apache.spark.rdd.PairRDDFunctions.saveAsHadoopDataset(PairRDDFunctions.scala:1089)
	at org.apache.spark.rdd.PairRDDFunctions.$anonfun$saveAsHadoopFile$4(PairRDDFunctions.scala:1062)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:410)
	at org.apache.spark.rdd.PairRDDFunctions.saveAsHadoopFile(PairRDDFunctions.scala:1027)
	at org.apache.spark.rdd.PairRDDFunctions.$anonfun$saveAsHadoopFile$3(PairRDDFunctions.scala:1009)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:410)
	at org.apache.spark.rdd.PairRDDFunctions.saveAsHadoopFile(PairRDDFunctions.scala:1008)
	at org.apache.spark.rdd.PairRDDFunctions.$anonfun$saveAsHadoopFile$2(PairRDDFunctions.scala:965)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:410)
	at org.apache.spark.rdd.PairRDDFunctions.saveAsHadoopFile(PairRDDFunctions.scala:963)
	at org.apache.spark.rdd.RDD.$anonfun$saveAsTextFile$2(RDD.scala:1623)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:410)
	at org.apache.spark.rdd.RDD.saveAsTextFile(RDD.scala:1623)
	at org.apache.spark.rdd.RDD.$anonfun$saveAsTextFile$1(RDD.scala:1609)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:410)
	at org.apache.spark.rdd.RDD.saveAsTextFile(RDD.scala:1609)
	at org.apache.spark.ml.util.DefaultParamsWriter$.saveMetadata(ReadWrite.scala:413)
	at org.apache.spark.ml.Pipeline$SharedReadWrite$.$anonfun$saveImpl$1(Pipeline.scala:250)
	at org.apache.spark.ml.Pipeline$SharedReadWrite$.$anonfun$saveImpl$1$adapted(Pipeline.scala:247)
	at org.apache.spark.ml.util.Instrumentation$.$anonfun$instrumented$1(Instrumentation.scala:191)
	at scala.util.Try$.apply(Try.scala:213)
	at org.apache.spark.ml.util.Instrumentation$.instrumented(Instrumentation.scala:191)
	at org.apache.spark.ml.Pipeline$SharedReadWrite$.saveImpl(Pipeline.scala:247)
	at org.apache.spark.ml.PipelineModel$PipelineModelWriter.saveImpl(Pipeline.scala:346)
	at org.apache.spark.ml.util.MLWriter.save(ReadWrite.scala:168)
	at org.apache.spark.ml.PipelineModel$PipelineModelWriter.super$save(Pipeline.scala:344)
	at org.apache.spark.ml.PipelineModel$PipelineModelWriter.$anonfun$save$4(Pipeline.scala:344)
	at org.apache.spark.ml.MLEvents.withSaveInstanceEvent(events.scala:174)
	at org.apache.spark.ml.MLEvents.withSaveInstanceEvent$(events.scala:169)
	at org.apache.spark.ml.util.Instrumentation.withSaveInstanceEvent(Instrumentation.scala:42)
	at org.apache.spark.ml.PipelineModel$PipelineModelWriter.$anonfun$save$3(Pipeline.scala:344)
	at org.apache.spark.ml.PipelineModel$PipelineModelWriter.$anonfun$save$3$adapted(Pipeline.scala:344)
	at org.apache.spark.ml.util.Instrumentation$.$anonfun$instrumented$1(Instrumentation.scala:191)
	at scala.util.Try$.apply(Try.scala:213)
	at org.apache.spark.ml.util.Instrumentation$.instrumented(Instrumentation.scala:191)
	at org.apache.spark.ml.PipelineModel$PipelineModelWriter.save(Pipeline.scala:344)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:374)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
	at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.UnsatisfiedLinkError: 'boolean org.apache.hadoop.io.nativeio.NativeIO$Windows.access0(java.lang.String, int)'
	at org.apache.hadoop.io.nativeio.NativeIO$Windows.access0(Native Method)
	at org.apache.hadoop.io.nativeio.NativeIO$Windows.access(NativeIO.java:793)
	at org.apache.hadoop.fs.FileUtil.canRead(FileUtil.java:1249)
	at org.apache.hadoop.fs.FileUtil.list(FileUtil.java:1454)
	at org.apache.hadoop.fs.RawLocalFileSystem.listStatus(RawLocalFileSystem.java:601)
	at org.apache.hadoop.fs.FileSystem.listStatus(FileSystem.java:1972)
	at org.apache.hadoop.fs.FileSystem.listStatus(FileSystem.java:2014)
	at org.apache.hadoop.fs.ChecksumFileSystem.listStatus(ChecksumFileSystem.java:761)
	at org.apache.hadoop.fs.FileSystem.listStatus(FileSystem.java:1972)
	at org.apache.hadoop.fs.FileSystem.listStatus(FileSystem.java:2014)
	at org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter.getAllCommittedTaskPaths(FileOutputCommitter.java:334)
	at org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter.commitJobInternal(FileOutputCommitter.java:404)
	at org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter.commitJob(FileOutputCommitter.java:377)
	at org.apache.hadoop.mapred.FileOutputCommitter.commitJob(FileOutputCommitter.java:136)
	at org.apache.hadoop.mapred.OutputCommitter.commitJob(OutputCommitter.java:291)
	at org.apache.spark.internal.io.HadoopMapReduceCommitProtocol.commitJob(HadoopMapReduceCommitProtocol.scala:192)
	at org.apache.spark.internal.io.SparkHadoopWriter$.$anonfun$write$3(SparkHadoopWriter.scala:100)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.util.Utils$.timeTakenMs(Utils.scala:552)
	at org.apache.spark.internal.io.SparkHadoopWriter$.write(SparkHadoopWriter.scala:100)
	... 68 more


Check directory contents

http://localhost:50070/explorer.html#/model-registry/titanic-survival-classifier/


<a id='8'></a>
## 8. Model Loading

In [61]:
from pyspark.ml import PipelineModel
savedModel = PipelineModel.load(modelPath)

Py4JJavaError: An error occurred while calling o2935.partitions.
: java.lang.UnsatisfiedLinkError: 'boolean org.apache.hadoop.io.nativeio.NativeIO$Windows.access0(java.lang.String, int)'
	at org.apache.hadoop.io.nativeio.NativeIO$Windows.access0(Native Method)
	at org.apache.hadoop.io.nativeio.NativeIO$Windows.access(NativeIO.java:793)
	at org.apache.hadoop.fs.FileUtil.canRead(FileUtil.java:1249)
	at org.apache.hadoop.fs.FileUtil.list(FileUtil.java:1454)
	at org.apache.hadoop.fs.RawLocalFileSystem.listStatus(RawLocalFileSystem.java:601)
	at org.apache.hadoop.fs.FileSystem.listStatus(FileSystem.java:1972)
	at org.apache.hadoop.fs.FileSystem.listStatus(FileSystem.java:2014)
	at org.apache.hadoop.fs.FileSystem$4.<init>(FileSystem.java:2180)
	at org.apache.hadoop.fs.FileSystem.listLocatedStatus(FileSystem.java:2179)
	at org.apache.hadoop.fs.ChecksumFileSystem.listLocatedStatus(ChecksumFileSystem.java:783)
	at org.apache.hadoop.mapred.FileInputFormat.singleThreadedListStatus(FileInputFormat.java:285)
	at org.apache.hadoop.mapred.FileInputFormat.listStatus(FileInputFormat.java:244)
	at org.apache.hadoop.mapred.FileInputFormat.getSplits(FileInputFormat.java:332)
	at org.apache.spark.rdd.HadoopRDD.getPartitions(HadoopRDD.scala:210)
	at org.apache.spark.rdd.RDD.$anonfun$partitions$2(RDD.scala:294)
	at scala.Option.getOrElse(Option.scala:189)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:290)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:49)
	at org.apache.spark.rdd.RDD.$anonfun$partitions$2(RDD.scala:294)
	at scala.Option.getOrElse(Option.scala:189)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:290)
	at org.apache.spark.api.java.JavaRDDLike.partitions(JavaRDDLike.scala:61)
	at org.apache.spark.api.java.JavaRDDLike.partitions$(JavaRDDLike.scala:61)
	at org.apache.spark.api.java.AbstractJavaRDDLike.partitions(JavaRDDLike.scala:45)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:374)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
	at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
	at java.base/java.lang.Thread.run(Thread.java:834)


In [65]:
predictions = savedModel.transform(testData)
predictions.select("features", "Survived", "prediction").limit(200).toPandas()

NameError: name 'savedModel' is not defined

<p style="color:white">https://www.kaggle.com/startupsci/titanic-data-science-solutions</p>

<a id='9'></a>
## 9. Tear Down

Once we complete the the lab we can stop all the services

<a id='9.1'></a>
### 9.1 Stop Hadoop

Stops Hadoop
Open a terminal and execute
```sh
hadoop-stop.sh
```